### DISCUSSION TITLE
Cannot find Microsoft.WindowsAppSDK.Bootstrap.dll

### DISCUSSION THREAD

**Discussion by capegreg** [microsoft/WindowsAppSDK#1320] [Q&A]
In **Tutorial: Build and deploy an unpackaged app that uses the Windows App SDK**, I'm getting an error unable to find Microsoft.WindowsAppSDK.Bootstrap.dll. I followed the tutorial exactly. 

Tutorial link
[https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=csharp-dotnet](url)

Error
`System.DllNotFoundException: 'Unable to load DLL 'Microsoft.WindowsAppSDK.Bootstrap.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)'`

I installed Windows App SDK (Experimental) through the modify installation for Visual Studio 2019 Individual components. The project compiles.

Reference code from tutorial

```
MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion);
args: 65536, 'experimental1', minVersion Build=0 Major=0 Minor=0 Revision=0

[DllImport("Microsoft.WindowsAppSDK.Bootstrap.dll", CharSet = CharSet.Unicode)]
private static extern int MddBootstrapInitialize(uint majorMinorVersion, string versionTag, PackageVersion packageVersion);

```

My dev environment
Microsoft Visual Studio Professional 2019
Version 16.11.2
WindowsAppSDK, version 1.0 Experimental

**Comment by DrusTheAxe** [comment]
What type of project are you building? There's a bug in the NuGet that the bootstrap DLL wasn't getting copied to a C# project's $(OutDir) post-build. Working on a fix for upcoming release but until then you'll need to manually copy the bootstrap dll from the NuGet to your $(OutDir) next to your app

Sorry for the inconvenience and thank you for your patience

**Reply by capegreg** [reply]
Thank you for your assistance. I can see the dll going to output now, but next issue is fatal error on it. Is there anything else I can document here to help troubleshoot it?

I created a brand new c++ x86 project following the instructions here [https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp)

**In the Post-Build Event was** 
xcopy /y "$(MSBuildThisFileDirectory)..\..\runtimes\lib\native\$(_WindowsAppSDKFoundationPlatform)\Microsoft.WindowsAppSDK.Bootstrap.dll" "$(OutDir)"

**Building the project shows the Bootstrap dll copied here**
C:\source\repos\DynamicDependenciesTest\Debug\Microsoft.WindowsAppSDK.Bootstrap.dll

**Running project gets unhandled error on this line**
const HRESULT hr{ MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion) };

Unhandled exception at 0x7C87120A (Microsoft.WindowsAppSDK.Bootstrap.dll) in DynamicDependenciesTest.exe: Fatal program exit requested.

**Reply by DrusTheAxe** [reply]
Are you running elevated? We don't support elevation today due to a mix of implementation details and Windows limitations. This issue's tracked via [Dynamic Dependencies doesn't support Elevation #567](https://github.com/microsoft/WindowsAppSDK/issues/567). When first discovered this led to cryptic errors (ERROR_NOT_FOUND or ERROR_CLASS_DB_NOTREG, depending) so an explicit trap was put in to detect if running elevated and FailFast. If this is your problem and you run under the debugger you'll see a message in the debugger's output window.

If you run under the debugger and it's not the explicit elevation detect/fail it would be interesting to see the callstack when the exception occurs (I can't think of any reason why this might occur besides the explicit elevation check/failfast)

**Reply by capegreg** [reply]
Yes. I now tried without elevated and the dll loaded and the code ran but returned this error. 

'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\source\repos\DynamicDependenciesTest\Debug\DynamicDependenciesTest.exe'. Symbols loaded.
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\ntdll.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\kernel32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\KernelBase.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\source\repos\DynamicDependenciesTest\Debug\Microsoft.WindowsAppSDK.Bootstrap.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\advapi32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\msvcrt.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\sechost.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\rpcrt4.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\ole32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\ucrtbase.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\combase.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\gdi32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\win32u.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\gdi32full.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\msvcp_win.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\user32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\oleaut32.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\msvcp140d.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\vcruntime140d.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\ucrtbased.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\imm32.dll'. 
The thread 0x4480 has exited with code 0 (0x0).
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\kernel.appcore.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\bcryptprimitives.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\clbcatq.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\AppExtension.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\SHCore.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\propsys.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\twinapi.appcore.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\AppXDeploymentClient.dll'. 
'DynamicDependenciesTest.exe' (Win32): Loaded 'C:\Windows\SysWOW64\Windows.StateRepositoryPS.dll'. 
D:\a\_work\1\s\dev\WindowsAppSDK_BootstrapDLL\MddBootstrap.cpp(339)\Microsoft.WindowsAppSDK.Bootstrap.dll!7B9E2719: (caller: 7B9E12D0) Exception(1) tid(f8c) 80070491 There was no match for the specified key in the index.
    Msg:[AppExtension.Name=microsoft.winappsdk.ddlm-1.0-x8-e1, Major=1, Minor=0, Tag=experimental1, MinVersion=0.0.0.0] 
Exception thrown at 0x75F8B512 in DynamicDependenciesTest.exe: Microsoft C++ exception: wil::ResultException at memory location 0x001DD45C.
Exception thrown at 0x75F8B512 in DynamicDependenciesTest.exe: Microsoft C++ exception: [rethrow] at memory location 0x00000000.
D:\a\_work\1\s\dev\WindowsAppSDK_BootstrapDLL\MddBootstrap.cpp(99)\Microsoft.WindowsAppSDK.Bootstrap.dll!7B9E1770: (caller: 002925DB) ReturnHr(1) tid(f8c) 80070491 There was no match for the specified key in the index.
    Msg:[D:\a\_work\1\s\dev\WindowsAppSDK_BootstrapDLL\MddBootstrap.cpp(339)\Microsoft.WindowsAppSDK.Bootstrap.dll!7B9E2719: (caller: 7B9E12D0) Exception(1) tid(f8c) 80070491 There was no match for the specified key in the index.
    Msg:[AppExtension.Name=microsoft.winappsdk.ddlm-1.0-x8-e1, Major=1, Minor=0, Tag=experimental1, MinVersion=0.0.0.0] 
] 
The thread 0x3afc has exited with code -2147023727 (0x80070491).
The thread 0x4408 has exited with code -2147023727 (0x80070491).
The thread 0x3fc4 has exited with code -2147023727 (0x80070491).
The thread 0x2920 has exited with code -2147023727 (0x80070491).
The thread 0x4508 has exited with code -2147023727 (0x80070491).
The thread 0x4284 has exited with code -2147023727 (0x80070491).
The thread 0x4108 has exited with code -2147023727 (0x80070491).
The program '[15784] DynamicDependenciesTest.exe' has exited with code -2147023727 (0x80070491).

**Reply by capegreg** [reply]
Update

I am able to successfully run the project [Tutorial: Build and deploy an unpackaged app that uses the Windows App SDKt](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=csharp-dotnet).

**Comment by capegreg** [comment]
Thanks, Howard. Yes, I am building a c# app. Do you have a link that shows how to do this?

**Reply by DrusTheAxe** [reply]
Manually, it's just a file copy via command line, Explorer or other tools or scripting.

Automatically? Sorry, no, I'm not familiar with a specific solution off hand. I mean, sure, it's possible, depending on your toolchain. You mentioned building a C# app, but didn't say how. If you use Visual Studio it's possible to add a `<PostBuildEvent>` to your project to copy the dll from the winappsdk's NuGet e.g. 

```
xcopy.exe ...\packages\....\Microsoft.WindowsAppSDK.Bootstrap.1.0-experimental1.dll $(OutDir)
```

There's probably more elegant ways to do it via MSBuild tags, or other ways if you're not using VS as your build environment.

**Reply by riverar** [reply]
@capegreg The Windows App SDK NuGet package will automatically handle deploying the bootstrap dll with your app. (Assuming you're using 1.0-experimental1.)

Here's a scratch unpackaged (unsupported) application that uses the WAS NuGet package for your reference
https://github.com/riverar/mux-repros/tree/master/unpackaged/scratch

**Reply by riverar** [reply]
For everyone's reference, the relevant lines from `Microsoft.WindowsAppSDK.Bootstrap.targets`:
https://github.com/microsoft/WindowsAppSDK/blob/aa4e265fa6cd02154981e75a288e85b39ab22357/build/NuSpecs/Microsoft.WindowsAppSDK.Bootstrap.targets#L10-L12

**Comment by bgun75** [comment]
You can try to copy the Microsoft.WindowsAppSDK.Bootstrap.dll file from your binary directory to c:\Windows\system32\.

**Reply by riverar** [reply]
No need to do that, thankfully.
